ml_metadata/tools/mlmd_bench/thread_runner.cc

```diff
@@ -0,0 +1,120 @@
+    {
+      // Create a thread pool for multi-thread execution.
+      tensorflow::thread::ThreadPool pool(tensorflow::Env::Default(),
+                                          "mlmd_bench", num_threads_);
+      // `approx_total_done` is used for reporting progress along the way.
+      int64 approx_total_done = 0;
+      for (int64 thread_index = 0; thread_index < num_threads_;
+           ++thread_index) {
+        int64 work_items_start_index = op_per_thread * thread_index;
+        ThreadStats& curr_thread_stats = thread_stats_list[thread_index];
+        pool.Schedule([this, op_per_thread, workload, work_items_start_index,
+                       &curr_thread_stats, &approx_total_done]() {
+          // Each thread uses a different MLMD client instance to talk to
+          // the same back-end.
+          std::unique_ptr<MetadataStore> store;
+          tensorflow::Status store_status =
+              CreateMetadataStore(mlmd_config_, &store);
+          // Handles abort issues for concurrent creating `store` of the db.
+          while (!store_status.ok()) {
+            store_status = CreateMetadataStore(mlmd_config_, &store);
+            // If the error is not Abort error, break the current process.
+            if (!store_status.ok() &&
+                store_status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(store_status);
+            }
+          }
+          curr_thread_stats.Start();
+          // Executes the current workload by the specified
+          // `work_items_index`.
+          int64 work_items_index = work_items_start_index;
+          while (work_items_index < work_items_start_index + op_per_thread) {
+            // Each operation has a op_stats.
+            OpStats op_stats;
+            tensorflow::Status status =
+                workload->RunOp(work_items_index, store.get(), op_stats);
+            // If the error is not Abort error, break the current process.
+            if (!status.ok() && status.code() != tensorflow::error::ABORTED) {
+              TF_CHECK_OK(status);
+            }
+            // Handles abort issues for concurrent writing to the db.
+            if (!status.ok()) {
+              continue;
+            }
+            work_items_index++;
+            approx_total_done++;
+            // Updates the current thread stats using the `op_stats`.
+            curr_thread_stats.Update(op_stats, approx_total_done);
+          }
+          curr_thread_stats.Stop();
+        });
+      }
+    }
+    TF_RETURN_IF_ERROR(workload->TearDown());
+    // Merges all the thread stats of the current workload.
+    for (int64 i = 1; i < num_threads_; ++i) {
+      thread_stats_list[0].Merge(thread_stats_list[i]);
+    }
+    // Reports the metrics of interests.
+    // TODO(briansong) Return the report as a summary proto and write it to a
+    // file.
+    thread_stats_list[0].Report(workload->GetName());
+  }
+  return tensorflow::Status::OK();
+}
```

Comment:
consider to move some part to internal functions, currently this method is very long.

Reply:
My bad! Fixed! Will be every carefully about this in the future.